casper/src/main/scala/coop/rchain/casper/Casper.scala

```diff
@@ -161,8 +162,31 @@ sealed abstract class MultiParentCasperInstances {
                            s"Block ${PrettyPrinter.buildString(b.blockHash)} is already being processed by another thread.")
                          .map(_ => BlockStatus.processing)
                      case Right((_, true)) =>
-                       internalAddBlock(b).flatMap(status =>
-                         Capture[F].capture { processingBlocks.update(_ - b.blockHash); status })
+                       for {
+                         validateFields <- Validate.formatOfFields[F](b)
+                         result <- validateFields match {
+                                    case Left(invalid) => Monad[F].pure[BlockStatus](invalid)
+                                    case Right(blockSafe) =>
+                                      for {
+                                        isOldBlock <- contains(blockSafe)
+                                        result <- if (isOldBlock) {
+                                                   for {
+                                                     _ <- Log[F].info(
+                                                           s"Received block ${PrettyPrinter
+                                                             .buildString(b.blockHash)} again.")
+                                                     result <- Monad[F].pure[BlockStatus](
+                                                                InvalidUnslashableBlock)
```

Comment:
I don't really like that a block we already have is considered invalid. Maybe make a new status to cover this case.

Reply:
Can we extract this into a new method? It has become hard to read for me.